### PR TITLE
EnumValue stemming/naming improvements

### LINKED ISSUE #621: Enums improperly stemmed and conflicting
A bug I encountered while trying out a recent version of the code on some Google-internal protos. Consider the following enum:

```proto
enum FooBar {
  FOO_BAR_BAZ = 0;
  QUUX = 1;
  BAZ = 2;
}
```

Our current stemming logic produces the following Swift enum cases:

```swift
baz = 0
quux = 1
baz = 2
```

Obviously, this doesn't compile. _(Why would someone design an enum like that to begin with?_ 🤷‍♂️  _But that ship has already sailed.)_

So, our stemming logic needs to change. Some options:

* Instead of stemming each case in isolation, stem in the context of all the cases and check for conflicts.
* Always stem "all-or-nothing", the logic we had before.
* Something else.

### PR DESCRIPTION
**Pull request by thomasvl**
Fixes #621

### CODE REVIEW

File: Sources/PluginLibrary/Descriptor.swift [hunk 1]
```diff
@@ -285,6 +286,9 @@ public final class EnumValueDescriptor {
     self.index = index
     self.enumType = enumType
     aliasOf = aliasing
+
+    let fullName = "\(enumType.fullName).\(proto.name)"
```
Comment (allevato) [inline Sources/PluginLibrary/Descriptor.swift]: Does this work better as a `lazy` property (or do you still have performance concerns there)?

File: Sources/PluginLibrary/NamingUtils.swift [hunk 2]
```diff
@@ -326,63 +326,61 @@ public enum NamingUtils {
...
+    }
+
+    /// Strip the prefix and return the result, or return nil if it can't
+    /// be stripped.
+    func strip(from: String) -> String? {
+      var prefixIndex = prefixChars.startIndex
+      let prefixEnd = prefixChars.endIndex
+
+      let fromChars = from.lowercased().unicodeScalars
+      precondition(fromChars.count == from.lengthOfBytes(using: .ascii))
```
Comment (allevato) [inline Sources/PluginLibrary/NamingUtils.swift]: This is just a sanity check that the case name we got was ASCII-only, right? Did you have any issues where this wasn't the case?

File: Sources/PluginLibrary/NamingUtils.swift [hunk 3]
```diff
@@ -326,63 +326,61 @@ public enum NamingUtils {
...
+      var fromIndex = fromChars.startIndex
+      let fromEnd = fromChars.endIndex
+
+      while (prefixIndex != prefixEnd) {
+        if (fromIndex == fromEnd) {
```
Comment (allevato) [inline Sources/PluginLibrary/NamingUtils.swift]: Nit: I would personally use `guard`s for many of these early-return checks, but I don't have a strong opinion about it.

File: Sources/PluginLibrary/SwiftProtobufNamer.swift [hunk 4]
```diff
@@ -81,18 +82,84 @@ public final class SwiftProtobufNamer {
...
-    let baseName = enumValue.name
-    if let stripped = NamingUtils.strip(protoPrefix: enumValue.enumType.name, from: baseName) {
-      let camelCased = NamingUtils.toLowerCamelCase(stripped)
-      if isValidSwiftIdentifier(camelCased) {
-        return NamingUtils.sanitize(enumCaseName: camelCased)
+  /// Compute the short names to use for the values of this enum.
+  private func computeRelativeNames(enum e: EnumDescriptor) {
+    let stripper = NamingUtils.PrefixStripper(prefix: e.name)
+
+    /// Determine the initial canidate name for the name before
```
Comment (allevato) [inline Sources/PluginLibrary/SwiftProtobufNamer.swift]: Nit: `//`

File: Sources/PluginLibrary/SwiftProtobufNamer.swift [hunk 5]
```diff
@@ -81,18 +82,84 @@ public final class SwiftProtobufNamer {
...
+      if let stripped = stripper.strip(from: baseName) {
+        let camelCased = NamingUtils.toLowerCamelCase(stripped)
+        if isValidSwiftIdentifier(camelCased) {
+          return camelCased
+        }
+      }
+      return NamingUtils.toLowerCamelCase(baseName)
+    }
+
+    // Bucked based on canidate names to check for duplicates.
```
Comment (allevato) [inline Sources/PluginLibrary/SwiftProtobufNamer.swift]: `Bucketed`? and `candidate`

File: Sources/PluginLibrary/SwiftProtobufNamer.swift [hunk 6]
```diff
@@ -81,18 +82,84 @@ public final class SwiftProtobufNamer {
...
+    for (camelCased, enumValues) in canidates {
+      // If there is only one, sanitize and cache it.
+      guard enumValues.count > 1 else {
+        enumValueRelativeNameCache[enumValues.first!.fullName] =
+          NamingUtils.sanitize(enumCaseName: camelCased)
+        continue
+      }
+
+      // There are two possible cases:
+      // 1. There is the main entry and then all aliase for it that
```
Comment (allevato) [inline Sources/PluginLibrary/SwiftProtobufNamer.swift]: aliases

File: Sources/PluginLibrary/Descriptor.swift (hunk 1 above)
Comment (thomasvl) [inline Sources/PluginLibrary/Descriptor.swift]: I've been trying to keep `lazy` to a minimum in here pending some performance analysis on generation, as there is an easy way to thread the plugin and generate files in parallel, but all of the lazy usage would have to go.

File: Sources/PluginLibrary/NamingUtils.swift (hunk 2 above)
Comment (thomasvl) [inline Sources/PluginLibrary/NamingUtils.swift]: Yea, was just a sanity check, and I'm just assuming protoc will do this instead.

File: Sources/PluginLibrary/NamingUtils.swift (hunk 3 above)
Comment (thomasvl) [inline Sources/PluginLibrary/NamingUtils.swift]: So I've been trying to figure out a pattern for `guard` vs `if`, the two possible reasons I've come up with is if a `let` is needed for the rest of the code, then `guard` seems to make more sense.  The other is when the logic gate reads better with an `else` meaning use the guard.  But yea, it is hard to say when one is better in a lot of cases.  ☹️

File: Sources/PluginLibrary/SwiftProtobufNamer.swift (hunk 4 above)
Comment (thomasvl) [inline Sources/PluginLibrary/SwiftProtobufNamer.swift]: Even though it documents the nested function?

File: Sources/PluginLibrary/SwiftProtobufNamer.swift (hunk 5 above)
Comment (thomasvl) [inline Sources/PluginLibrary/SwiftProtobufNamer.swift]: Done.

File: Sources/PluginLibrary/SwiftProtobufNamer.swift (hunk 6 above)
Comment (thomasvl) [inline Sources/PluginLibrary/SwiftProtobufNamer.swift]: Done.

File: Sources/PluginLibrary/SwiftProtobufNamer.swift (hunk 4 above)
Comment (allevato) [inline Sources/PluginLibrary/SwiftProtobufNamer.swift]: Woops, I didn't even realize that was a nested function—I thought it was just a comment inside this function. Nvm.

File: Sources/PluginLibrary/NamingUtils.swift (hunk 3 above)
Comment (allevato) [inline Sources/PluginLibrary/NamingUtils.swift]: To me, `guard` is a nice visual indicator of "this checks a condition that demands an early exit from this scope (or more rarely, loop continue)"; it highlights special cases well whether or not there's a binding involved. (Though it absolutely makes the use of that binding easier as well by avoiding extra levels of nesting.)

**Pull request merged**